Mathlib/Tactic/Linter/TextBased.lean

```diff
@@ -0,0 +1,203 @@
+        | "ERR_NUM_LIN" =>
+          -- Parse the error message in the script. `none` indicates invalid input.
+          match (error_message.get? 0, error_message.get? 3) with
+          | (some limit, some current) =>
+            match (String.toNat? limit, String.toNat? current) with
+            | (some size_limit, some current_size) =>
+              some (StyleError.fileTooLong current_size size_limit)
+            | _ => none
+          | _ => none
+        | _ => none
+      -- Omit the line number, as we don't use it anyway.
+      return (err.map fun e ↦ (ErrorContext.mk e 0 path))
+    -- XXX: print an error on any lines which don't match the particular format.
+    | _ => -- IO.println s!"Invalid input file: line {line} doesn't match any particular format"
+      none
+
+/-- Parse all style exceptions for a line of input.
+Return an array of all exceptions which could be parsed: invalid input is ignored. -/
+def parseStyleExceptions (lines : Array String) : Array ErrorContext := Id.run do
+  Array.filterMap (fun line ↦ parse?_errorContext line) lines
+
+/-- Print information about all errors encountered to standard output. -/
+def formatErrors (errors : Array ErrorContext) : IO Unit := do
+  for e in errors do
+    IO.println (outputMessage e)
+
+/-- Core logic of a text based linter: given a collection of lines,
+return an array of all style errors with line numbers. -/
+abbrev TextbasedLinter := Array String → Array (StyleError × ℕ)
+
+/-! Definitions of the actual text-based linters. -/
+section
+
+/-- Whether a collection of lines consists *only* of imports, blank lines and single-line comments.
+In practice, this means it's an imports-only file and exempt from almost all linting. -/
+def isImportsOnlyFile (lines : Array String) : Bool :=
+  -- The Python version also excluded multi-line comments: for all files generated by `mk_all`,
+  -- this is in fact not necessary.
+  lines.all (fun line ↦ line.startsWith "import " || line == "" || line.startsWith "-- ")
+
+/-- Error if a collection of lines is too large. "Too large" means more than 1500 lines
+**and** longer than an optional previous limit.
+If the file is too large, return a matching `StyleError`, which includes a new size limit
+(which is somewhat larger than the current size). -/
+def checkFileLength (lines : Array String) (existing_limit : Option ℕ) : Option StyleError :=
+  Id.run do
+  if lines.size > 1500 then
+    let is_larger : Bool := match existing_limit with
+    | some mark => lines.size > mark
+    | none => true
+    if is_larger then
+      -- We add about 200 lines of slack to the current file size:
+      -- small PRs will be unperturbed, but sufficiently large PRs will get nudged towards
+      -- splitting up this file.
+      return some (StyleError.fileTooLong lines.size ((Nat.div lines.size 100) * 100 + 200))
+  none
+
+end
+
+/-- All text-based linters registered in this file. -/
+def allLinters : Array TextbasedLinter := Array.mk
```

Comment:
This list is currently empty (but will be filled when other linters get ported): I'm happy to remove this for now, if so desired.